libs/user_extensions/src/TTAlpsDimuonsSelections.cpp

```diff
@@ -0,0 +1,184 @@
+#include "TLorentzVector.h"
+
+using namespace std;
+
+TTAlpsDimuonSelections::TTAlpsDimuonSelections(){
+
+  auto &config = ConfigManager::GetInstance();
+
+  string selection = "SR";
+  try {
+    config.GetValue("dimuonSelection", selection);
+  } catch (const Exception &e) {
+    warn() << "Couldn't read dimuonSelection from config file - will use SR selection for LooseMuons" << endl;
+  }
+  dimuonSelection = selection;
+
+  try {
+    config.GetMap(dimuonSelection+"BaseCuts", dimuonVertexBaseCuts);
+  } catch (const Exception &e) {
+    info() << "Couldn't read " << dimuonSelection << "BaseCuts from config file - is needed for GoodLooseMuonVertex collections" << endl;
+  }
+  try {
+    config.GetMap(dimuonSelection+"PATCuts", dimuonVertexPATCuts);
+  } catch (const Exception &e) {
+    info() << "Couldn't read " << dimuonSelection << "PATCuts from config file - is needed for GoodLooseMuonVertex collections" << endl;
+  }
+  try {
+    config.GetMap(dimuonSelection+"PATDSACuts", dimuonVertexPATDSACuts);
+  } catch (const Exception &e) {
+    info() << "Couldn't read " << dimuonSelection << "PATDSACuts from config file - is needed for GoodLooseMuonVertex collections" << endl;
+  }
+  try {
+    config.GetMap(dimuonSelection+"DSACuts", dimuonVertexDSACuts);
+  } catch (const Exception &e) {
+    info() << "Couldn't read " << dimuonSelection << "DSACuts from config file - is needed for GoodLooseMuonVertex collections" << endl;
+  }
+
+  using namespace std::placeholders;
+  PassesCutsMap = {
+    {"LLPnanoAODVertexCuts", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesLLPnanoAODVertexCuts(v); }},
+    {"InvMassCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesInvariantMassCuts(v); }},
+    {"ChargeCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesChargeCut(v); }},
+    {"DisplacedIsoCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesDisplacedIsolationCut(v); }},
+    {"HitsInFrontOfVertexCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesHitsInFrontOfVertexCut(v); }},
+    {"DPhiMuonpTLxyCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesDPhiBetweenMuonpTAndLxyCut(v); }},
+    {"DCACut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesDCACut(v); }},
+    {"Chi2Cut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesChi2Cut(v); }},
+    {"VxyCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesVxyCut(v); }},
+    {"CollinearityCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesCollinearityAngleCut(v); }},
+    {"DeltaEtaCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesDeltaEtaCut(v); }},
+    {"DeltaPhiCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesDeltaPhiCut(v); }},
+    {"DeltaRCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesDeltaRCut(v); }},
+    {"DeltaPixelHitsCut", [this](std::shared_ptr<NanoDimuonVertex> v) { return PassesDeltaPixelHits(v); }}
+  };
+}
```

Comment:
So, I thought about this for a while and talked to Dr. GPT, but I don't think there is a good solution for this in C++.  The only alternative I could think of would be to directly define the cuts when creating this map, which reduces a bit the number of lines, and we don't have to define them in the header at all. You also don't need to call the selections one by one from the main. There are some drawbacks to this solution, too... Please have a look at the example below and let's discuss if we want to go this way or not:

```cpp
class TTAlpsDimuonSelections {
 public:
  TTAlpsDimuonSelections() {
    // read this from the config
    cuts_to_apply = {"vertex", "invariant_mass"};

    // define the possible selections directly here, using string keys that match the config
    selections["vertex"] = [this](float v) {
      if (v > 0) return true;
      return false;
    };

    selections["invariant_mass"] = [this](float v) {
      if (v > 1.0) return true;
      return false;
    };
  }
  ~TTAlpsDimuonSelections() {}

  bool PassesCuts(float v) {
    for (auto& method : cuts_to_apply) {
      if (!selections[method](v)) return false;
    }
    return true;
  }

 private:
  map<string, function<bool(float)>> selections;
  vector<string> cuts_to_apply;
};

int main() {
  TTAlpsDimuonSelections ttalps;
  bool passesAllCuts = ttalps.PassesCuts(2.0);
  cout << "passesAllCuts: " << passesAllCuts << endl;
  return 0;
}
```

Reply:
Hm the name of the strings are still kind of hard-coded here so it wouldn't make much difference. I think this makes it harder to follow as a user so I would prefer to do it with defined functions as usual, that way we could also call the functions for some other use if needed in the future

Reply:
But I'll make all strings be the exact same name as the strings now, it wasn't before